cmd/kueue/main.go

```diff
@@ -262,69 +253,38 @@ func setupControllers(mgr ctrl.Manager, cCache *cache.Cache, queues *queue.Manag
 		setupLog.Error(err, "Unable to create webhook", "webhook", failedWebhook)
 		os.Exit(1)
 	}
 
 	opts := []jobframework.Option{
-		jobframework.WithManageJobsWithoutQueueName(manageJobsWithoutQueueName),
-		jobframework.WithWaitForPodsReady(waitForPodsReady(cfg)),
+		jobframework.WithManageJobsWithoutQueueName(cfg.ManageJobsWithoutQueueName),
+		jobframework.WithWaitForPodsReady(config.IsWaitForPodsReadyEnabled(cfg.WaitForPodsReady)),
 		jobframework.WithKubeServerVersion(serverVersionFetcher),
+		jobframework.WithEnabledFrameworks(config.EnabledFrameworks(cfg.Integrations)),
+		jobframework.WithManagerName(constants.KueueName),
 	}
-	err := jobframework.ForEachIntegration(func(name string, cb jobframework.IntegrationCallbacks) error {
-		log := setupLog.WithValues("jobFrameworkName", name)
-		if isFrameworkEnabled(cfg, name) {
-			gvk, err := apiutil.GVKForObject(cb.JobType, mgr.GetScheme())
-			if err != nil {
-				return err
-			}
-			if _, err = mgr.GetRESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version); err != nil {
-				if !meta.IsNoMatchError(err) {
-					return err
-				}
-				log.Info("No matching API in the server for job framework, skipped setup of controller and webhook")
-			} else {
-				if err = cb.NewReconciler(
-					mgr.GetClient(),
-					mgr.GetEventRecorderFor(fmt.Sprintf("%s-%s-controller", name, constants.KueueName)),
-					opts...,
-				).SetupWithManager(mgr); err != nil {
-					log.Error(err, "Unable to create controller")
-					return err
-				}
-				if name == "pod" {
-					v := serverVersionFetcher.GetServerVersion()
-					if v.String() == "" || v.LessThan(kubeversion.KubeVersion1_27) {
-						setupLog.Error(errPodIntegration,
-							"Failed to configure reconcilers",
-							"kubernetesVersion", v)
-						os.Exit(1)
-					}
-
-					opts = append(
-						opts,
-						jobframework.WithPodNamespaceSelector(cfg.Integrations.PodOptions.NamespaceSelector),
-						jobframework.WithPodSelector(cfg.Integrations.PodOptions.PodSelector),
-					)
-				}
-				if err = cb.SetupWebhook(mgr, opts...); err != nil {
-					log.Error(err, "Unable to create webhook")
-					return err
-				}
-				log.Info("Set up controller and webhook for job framework")
-				return nil
-			}
+	err := jobframework.SetupControllers(mgr, setupLog, func(fwkName string, opts ...jobframework.Option) ([]jobframework.Option, error) {
+		options := jobframework.DefaultOptions
+		for _, opt := range opts {
+			opt(&options)
 		}
-		if err := noop.SetupWebhook(mgr, cb.JobType); err != nil {
-			log.Error(err, "Unable to create noop webhook")
-			return err
+		if fwkName == "pod" {
+			v := options.KubeServerVersion.GetServerVersion()
+			if v.String() == "" || v.LessThan(kubeversion.KubeVersion1_27) {
+				return opts, errPodIntegration
+			}
+			return append(
+				opts,
+				jobframework.WithPodNamespaceSelector(cfg.Integrations.PodOptions.NamespaceSelector),
+				jobframework.WithPodSelector(cfg.Integrations.PodOptions.PodSelector),
+			), nil
```

Comment:
I really don't like this... It might be better if `SetupWebhook` accepts an `any` argument that it can then cast to whatever struct it understands.

Then we could have something like: `jobframework.WithIntegrationOptions("pod", cfg.Integrations.PodOptions)`.

We did something similar to pass arguments to different plugins in kube-scheduler https://github.com/kubernetes/kubernetes/blob/0f817dcd6577b663f6cb22cd5080f14138e8b6ac/pkg/scheduler/framework/plugins/interpodaffinity/plugin.go#L73
https://github.com/kubernetes/kubernetes/blob/0f817dcd6577b663f6cb22cd5080f14138e8b6ac/pkg/scheduler/framework/plugins/interpodaffinity/plugin.go#L95

Reply:
It sounds reasonable, but I think that working on its refactoring in another PR (before or after) would be better. 
WDYT?

Reply:
I'm ok with taking this as a prerequisite. Let's hold this PR until we finalize this refactoring.
/hold

Reply:
@alculquicondor I opened a PR here: #1637